lib/libzfs/libzfs_util.c

```diff
@@ -2143,3 +2143,50 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
 {
 	return (zprop_iter_common(func, cb, show_all, ordered, type));
 }
+
+/*
+ * zfs_get_hole_count retrieves the number of holes (blocks which are
+ * zero-filled) in the specified file using the ZFS_IOC_COUNT_FILLED ioctl. It
+ * also optionally fetches the block size when bs is non-NULL. With hole count
+ * and block size the full space consumed by the holes of a file can be
+ * calculated.
+ *
+ * On success, zero is returned, the count argument is set to the
+ * number of holes, and the bs argument is set to the block size (if it is
```

Comment:
Out of curiosity, it isn't clear how the blocksize is useful here, though I guess it doesn't add significantly to the overhead?  This is already returned by `stat()` for the file, and it isn't like `blocksize * number_of_holes` is a useful value?  Hmm, maybe the confusion is that "the number of holes" is actually "the number of unallocated *blocks*" and not "the number of contiguous unallocated regions" (which is what I'd consider the traditional definition of "hole" is).

I guess in that case blocksize is useful, but the comment (and description of this feature in general) should be improved.  Rather than describe this as "number of holes", I'd write "number of unallocated (sparse) blocks inside a file".

In that case `FS_IOC_FIEMAP` with `fm_extent_count = 0` is unfortunately *not* a suitable replacement for this functionality.

Reply:
That's covered in the first part of the comment, but I'll reiterate it here too

```
/*
 * zfs_get_hole_count retrieves the number of holes (blocks which are
 * zero-filled) in the specified file using the ZFS_IOC_COUNT_FILLED ioctl.
```